File: app/models/order.py
from datetime import datetime, timezone
from enum import Enum
from typing import List, Optional
from beanie import Document, Indexed
from pydantic import BaseModel, Field
from app.schemas.location import GPSLocation


class OrderType(str, Enum):
    PRODUCT_ORDER = "product_order"
    PRINT_SERVICE = "print_service"
    PORTER_SERVICE = "porter_service"
    ASSIGNMENT_WRITER = "assignment_writer"


class OrderStatus(str, Enum):
    PENDING = "pending"
    ACCEPTED = "accepted"
    ASSIGNED = "assigned"
    DOCUMENT_PICKED_UP = "document_picked_up"
    OUT_FOR_DELIVERY = "out_for_delivery"
    DELIVERED = "delivered"
    CANCELLED = "cancelled"


class PaymentMethod(str, Enum):
    UPI = "upi"
    CASH = "cash"
    RAZORPAY = "razorpay"


class PaymentStatus(str, Enum):
    PENDING = "pending"
    PAID = "paid"
    CASH_ON_DELIVERY = "cash_on_delivery"
    FAILED = "failed"



class OrderItem(BaseModel):
    product_id: str
    product_name: str
    quantity: int = Field(1, ge=1)
    unit_price: float = Field(..., ge=0.0)
    subtotal: float = Field(..., ge=0.0)


class PrintServiceSpec(BaseModel):
    """Specification for Xerox, Printing & Binding services."""

    file_url: Optional[str] = Field(None, description="URL of uploaded document (PDF/Docx)")
    document_name: str = Field("Document", description="Name of document to print")
    is_physical_pickup: bool = Field(False, description="True if partner physically picks up hardcopy document from customer for Xerox")
    num_pages: int = Field(1, ge=1)
    num_copies: int = Field(1, ge=1)
    color_mode: str = Field("black_and_white", description="black_and_white or color")
    paper_size: str = Field("A4", description="A4, A3, Letter")
    is_double_sided: bool = False
    binding_type: str = Field("none", description="none, spiral, channel_file")
    special_instructions: Optional[str] = None


class AssignmentServiceSpec(BaseModel):
    """Specification for Handwritten Assignment Writer services."""

    file_url: Optional[str] = Field(None, description="URL of reference uploaded document/images")
    document_name: str = Field("Assignment", description="Name or topic of assignment")
    is_physical_pickup: bool = Field(False, description="True if partner physically picks up notebook/pages/hardcopy notes from customer")
    num_pages: int = Field(1, ge=1, description="Number of pages to be handwritten")
    paper_type: str = Field("a4_ruled", description="a4_ruled, a4_unruled, practical_sheet")
    binding_type: str = Field("none", description="none (loose paper), spiral (spiral file), channel_file (channel file)")
    ink_color: str = Field("blue", description="blue, black, blue_black, multicolor")
    special_instructions: Optional[str] = None


class PorterServiceSpec(BaseModel):
    """Specification for Porter Parcel Courier Service (strictly under 5 kg)."""

    item_description: str = Field(..., description="Description of item/parcel to deliver")
    weight_kg: float = Field(..., le=5.0, gt=0.0, description="Item weight must be 5.0 kg or less")
    pickup_address: str = Field(..., description="Full pickup street address")
    pickup_location: Optional[GPSLocation] = None
    drop_address: str = Field(..., description="Full drop-off street address")
    drop_location: Optional[GPSLocation] = None
    sender_phone: str
    receiver_phone: str
    notes: Optional[str] = None


class Order(Document):
    """MongoDB Document model for Quick-Commerce, Print/Xerox, Porter (< 5kg), and Assignment Writer orders."""

    order_id: Indexed(str, unique=True)
    customer_id: str
    partner_id: Optional[str] = None  # Delivery partner assigned
    order_type: OrderType
    status: OrderStatus = OrderStatus.PENDING

    # Specific payload sections
    items: List[OrderItem] = Field(default_factory=list)
    print_spec: Optional[PrintServiceSpec] = None
    porter_spec: Optional[PorterServiceSpec] = None
    assignment_spec: Optional[AssignmentServiceSpec] = None

    # Payment details
    payment_method: PaymentMethod = PaymentMethod.CASH
    payment_status: PaymentStatus = PaymentStatus.PENDING
    upi_transaction_id: Optional[str] = None
    razorpay_order_id: Optional[str] = None
    razorpay_payment_id: Optional[str] = None
    razorpay_signature: Optional[str] = None


    # Delivery information
    delivery_address: Optional[str] = None
    delivery_location: Optional[GPSLocation] = None
    customer_phone: Optional[str] = None

    # Ringing partner list (partners within 1KM radius notified)
    notified_partner_ids: List[str] = Field(default_factory=list)

    # Financial details
    items_total: float = Field(0.0, ge=0.0)
    delivery_fee: float = Field(0.0, ge=0.0)
    total_amount: float = Field(0.0, ge=0.0)

    # Customer Rating & Review details
    is_rated: bool = Field(False, description="Whether customer has rated this order")
    rating: Optional[float] = Field(None, ge=1.0, le=5.0, description="Customer star rating (1.0 - 5.0)")
    review: Optional[str] = Field(None, description="Customer feedback review text")

    created_at: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc)
    )
    updated_at: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc)
    )

    class Settings:
        name = "orders"
        indexes = [
            "customer_id",
            "partner_id",
            "status",
            [("customer_id", 1), ("created_at", -1)],
            [("partner_id", 1), ("status", 1)],
            [("status", 1), ("created_at", -1)],
        ]

    def touch(self):
        self.updated_at = datetime.now(timezone.utc)

File: src/models/Order.ts
import mongoose, { Schema, Document } from 'mongoose';
import { IGPSLocation } from './User';

export enum OrderType {
  PRODUCT_ORDER = 'product_order',
  PRINT_SERVICE = 'print_service',
  PORTER_SERVICE = 'porter_service',
  ASSIGNMENT_WRITER = 'assignment_writer',
}

export enum OrderStatus {
  PENDING = 'pending',
  ACCEPTED = 'accepted',
  ASSIGNED = 'assigned',
  DOCUMENT_PICKED_UP = 'document_picked_up',
  OUT_FOR_DELIVERY = 'out_for_delivery',
  DELIVERED = 'delivered',
  CANCELLED = 'cancelled',
}

export enum PaymentMethod {
  UPI = 'upi',
  CASH = 'cash',
  RAZORPAY = 'razorpay',
}

export enum PaymentStatus {
  PENDING = 'pending',
  PAID = 'paid',
  CASH_ON_DELIVERY = 'cash_on_delivery',
  FAILED = 'failed',
}

export interface IOrderItem {
  product_id: string;
  product_name: string;
  quantity: number;
  unit_price: number;
  subtotal: number;
}

export interface IPrintServiceSpec {
  file_url?: string;
  document_name: string;
  is_physical_pickup: boolean;
  num_pages: number;
  num_copies: number;
  color_mode: string;
  paper_size: string;
  is_double_sided: boolean;
  binding_type: string;
  special_instructions?: string;
}

export interface IAssignmentServiceSpec {
  file_url?: string;
  document_name: string;
  is_physical_pickup: boolean;
  num_pages: number;
  paper_type: string;
  binding_type: string;
  ink_color: string;
  special_instructions?: string;
}

export interface IPorterServiceSpec {
  item_description: string;
  weight_kg: number;
  pickup_address: string;
  pickup_location?: IGPSLocation;
  drop_address: string;
  drop_location?: IGPSLocation;
  sender_phone: string;
  receiver_phone: string;
  notes?: string;
}

export interface IOrder extends Document {
  order_id: string;
  customer_id: string;
  partner_id?: string;
  order_type: OrderType;
  status: OrderStatus;
  items: IOrderItem[];
  print_spec?: IPrintServiceSpec;
  porter_spec?: IPorterServiceSpec;
  assignment_spec?: IAssignmentServiceSpec;
  payment_method: PaymentMethod;
  payment_status: PaymentStatus;
  upi_transaction_id?: string;
  razorpay_order_id?: string;
  razorpay_payment_id?: string;
  razorpay_signature?: string;
  delivery_address?: string;
  delivery_location?: IGPSLocation;
  customer_phone?: string;
  notified_partner_ids: string[];
  items_total: number;
  delivery_fee: number;
  total_amount: number;
  is_rated: boolean;
  rating?: number;
  review?: string;
  created_at: Date;
  updated_at: Date;
  touch(): void;
}

const GPSLocationSchema = new Schema(
  {
    latitude: { type: Number, required: true },
    longitude: { type: Number, required: true },
    accuracy: { type: Number },
    timestamp: { type: Date, default: Date.now },
    address: { type: String },
  },
  { _id: false }
);

const OrderItemSchema = new Schema(
  {
    product_id: { type: String, required: true },
    product_name: { type: String, required: true },
    quantity: { type: Number, required: true, min: 1, default: 1 },
    unit_price: { type: Number, required: true, min: 0 },
    subtotal: { type: Number, required: true, min: 0 },
  },
  { _id: false }
);

const PrintServiceSpecSchema = new Schema(
  {
    file_url: { type: String },
    document_name: { type: String, default: 'Document' },
    is_physical_pickup: { type: Boolean, default: false },
    num_pages: { type: Number, default: 1, min: 1 },
    num_copies: { type: Number, default: 1, min: 1 },
    color_mode: { type: String, default: 'black_and_white' },
    paper_size: { type: String, default: 'A4' },
    is_double_sided: { type: Boolean, default: false },
    binding_type: { type: String, default: 'none' },
    special_instructions: { type: String },
  },
  { _id: false }
);

const AssignmentServiceSpecSchema = new Schema(
  {
    file_url: { type: String },
    document_name: { type: String, default: 'Assignment' },
    is_physical_pickup: { type: Boolean, default: false },
    num_pages: { type: Number, default: 1, min: 1 },
    paper_type: { type: String, default: 'a4_ruled' },
    binding_type: { type: String, default: 'none' },
    ink_color: { type: String, default: 'blue' },
    special_instructions: { type: String },
  },
  { _id: false }
);

const PorterServiceSpecSchema = new Schema(
  {
    item_description: { type: String, required: true },
    weight_kg: { type: Number, required: true, min: 0, max: 5 },
    pickup_address: { type: String, required: true },
    pickup_location: { type: GPSLocationSchema },
    drop_address: { type: String, required: true },
    drop_location: { type: GPSLocationSchema },
    sender_phone: { type: String, required: true },
    receiver_phone: { type: String, required: true },
    notes: { type: String },
  },
  { _id: false }
);

const OrderSchema = new Schema<IOrder>(
  {
    order_id: { type: String, required: true, unique: true, index: true },
    customer_id: { type: String, required: true, index: true },
    partner_id: { type: String, index: true },
    order_type: { type: String, enum: Object.values(OrderType), required: true },
    status: { type: String, enum: Object.values(OrderStatus), default: OrderStatus.PENDING, index: true },
    items: [OrderItemSchema],
    print_spec: { type: PrintServiceSpecSchema },
    porter_spec: { type: PorterServiceSpecSchema },
    assignment_spec: { type: AssignmentServiceSpecSchema },
    payment_method: { type: String, enum: Object.values(PaymentMethod), default: PaymentMethod.CASH },
    payment_status: { type: String, enum: Object.values(PaymentStatus), default: PaymentStatus.PENDING },
    upi_transaction_id: { type: String },
    razorpay_order_id: { type: String },
    razorpay_payment_id: { type: String },
    razorpay_signature: { type: String },
    delivery_address: { type: String },
    delivery_location: { type: GPSLocationSchema },
    customer_phone: { type: String },
    notified_partner_ids: [{ type: String }],
    items_total: { type: Number, default: 0, min: 0 },
    delivery_fee: { type: Number, default: 0, min: 0 },
    total_amount: { type: Number, default: 0, min: 0 },
    is_rated: { type: Boolean, default: false },
    rating: { type: Number, min: 1, max: 5 },
    review: { type: String },
  },
  {
    timestamps: { createdAt: 'created_at', updatedAt: 'updated_at' },
  }
);

OrderSchema.index({ customer_id: 1, created_at: -1 });
OrderSchema.index({ customer_id: 1, status: 1, created_at: -1 });
OrderSchema.index({ partner_id: 1, status: 1, created_at: -1 });
OrderSchema.index({ notified_partner_ids: 1, status: 1, created_at: -1 });
OrderSchema.index({ status: 1, created_at: -1 });

OrderSchema.methods.touch = function () {
  this.updated_at = new Date();
};

export const Order = mongoose.model<IOrder>('Order', OrderSchema, 'orders');
